sql/rls: apply SELECT policies only if RETURNING references columns

spilchen opened the pull request:
Previously, SELECT policies were always applied when a RETURNING clause was used in an INSERT or UPDATE statement. This behavior was overly restrictive, as the SELECT policies should only be enforced when the RETURNING clause actually references one or more columns from the table.

This change modifies the logic to inspect the RETURNING clause and apply SELECT policies only when column references are present.

Fixes #144951

Epic: CRDB-11724

Release note (bug fix): Fixed incorrect application of SELECT policies
to RETURNING clauses in INSERT and UPDATE when no table columns were
referenced.

mgartner reviewed (commented):
> Reviewed 1 of 8 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145890)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/opt/optbuilder/insert.go` line 777 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145890#-OPBVYuX66F8l8weHh42:-OPBVYuX66F8l8weHh43:binunt0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b73b6fce76a314acf40edbfd835a0acb3da8d26f/pkg/sql/opt/optbuilder/insert.go#L777)):*
> ```Go
> 		var colRefs opt.ColSet
> 		_, _ = mb.buildReturningScopes(returning, &colRefs)
> 		includeSelectPolicies = colRefs.Len() > 0
> ```

For `INSERT ... RETURNING` I think it is safe to simply check if `colRefs` is empty or not because the `RETURNING` clause can only reference columns in the target table. Can you add a note to explain this?

----

nit: use `colRefs.Empty()` which is more efficient than `colRefs.Len()`
___
*[`pkg/sql/opt/optbuilder/update.go` line 365 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145890#-OPBYwus2H5qj3fN5j-r:-OPBYwus2H5qj3fN5j-s:b94ar1n) ([raw file](https://github.com/cockroachdb/cockroach/blob/b73b6fce76a314acf40edbfd835a0acb3da8d26f/pkg/sql/opt/optbuilder/update.go#L365)):*
> ```Go
> 		// aren't already needed.
> 		if returning != nil && colRefs != nil && colRefs.Len() == 0 {
> 			_, _ = mb.buildReturningScopes(returning, colRefs)
> ```

I don't see any reason we need to re-build the scopes later on. I think we can unconditionally build the returning scopes up here, and then use those same scopes to build the returning projection at the bottom of this function.
___
*[`pkg/sql/opt/optbuilder/update.go` line 367 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145890#-OPBW4mfCVdlSKE8EXqe:-OPBW4mfCVdlSKE8EXqf:b-3jc65m) ([raw file](https://github.com/cockroachdb/cockroach/blob/b73b6fce76a314acf40edbfd835a0acb3da8d26f/pkg/sql/opt/optbuilder/update.go#L367)):*
> ```Go
> 			_, _ = mb.buildReturningScopes(returning, colRefs)
> 		}
> 		includeSelectPolicies = colRefs != nil && colRefs.Len() > 0
> ```

Similar to my comments in #145344, we need to do more than check if colRefs is non-empty. Example:

```sql
DROP TABLE IF EXISTS t;
DROP TABLE IF EXISTS t2;
DROP ROLE IF EXISTS alice;

CREATE ROLE alice;

CREATE TABLE t (
  k INT PRIMARY KEY,
  a INT
);

INSERT INTO t VALUES (1, 10);

GRANT SELECT, INSERT, UPDATE, DELETE ON t TO alice;

ALTER TABLE t ENABLE ROW LEVEL SECURITY;

CREATE POLICY select_policy_alice
ON t
FOR SELECT
TO alice
USING (a > 0);

CREATE POLICY update_policy_alice
ON t
FOR UPDATE
TO alice
USING (true);

CREATE TABLE t2 (k INT PRIMARY KEY, a INT);

INSERT INTO t2 VALUES (1, -1);

GRANT SELECT, INSERT, UPDATE, DELETE ON t2 TO alice;

SET ROLE alice;

UPDATE t SET a = -2 FROM t2 WHERE true RETURNING t2.k;
-- In Postgres, this update succeeds:
--  k
-- ---
--  1
-- (1 row)
--
-- UPDATE 1
--
-- In CRDB with #144951, the update errors.
-- ERROR: new row violates row-level security policy for table "t"
```
___
*[`pkg/sql/logictest/testdata/logic_test/row_level_security` line 737 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145890#-OPBa8N5BIiQKR8uZD1t:-OPBa8N5BIiQKR8uZD1u:bu77ano) ([raw file](https://github.com/cockroachdb/cockroach/blob/b73b6fce76a314acf40edbfd835a0acb3da8d26f/pkg/sql/logictest/testdata/logic_test/row_level_security#L737)):*
> ```
> 
> statement error pq: new row violates row-level security policy for table "ins"
> INSERT INTO ins VALUES (7) RETURNING (select ins.c1);
> ```

Can you add a success case where the RETURNING clause has a subquery referencing another table, e.g., `INSERT INTO ins VALUES (7) RETURNING (SELECT k FROM t2);`? It looks like it works fine.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145890)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @spilchen)

___
*[`pkg/sql/opt/optbuilder/update.go` line 365 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145890#-OPBYwus2H5qj3fN5j-r:-OPGTeOB3cB7iYFsxh4_:b-2kson9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b73b6fce76a314acf40edbfd835a0acb3da8d26f/pkg/sql/opt/optbuilder/update.go#L365)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I don't see any reason we need to re-build the scopes later on. I think we can unconditionally build the returning scopes up here, and then use those same scopes to build the returning projection at the bottom of this function.
</blockquote></details>

Ah, got it—I initially thought the order mattered. I ended up creating the scopes unconditionally at the start and reusing them when projecting the RETURNING columns at the end, though I did need to update the scope's `expr`.
___
*[`pkg/sql/opt/optbuilder/update.go` line 367 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145890#-OPBW4mfCVdlSKE8EXqe:-OPGo1qW-yDg0FhOVOzL:b-4u87l8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b73b6fce76a314acf40edbfd835a0acb3da8d26f/pkg/sql/opt/optbuilder/update.go#L367)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Similar to my comments in #145344, we need to do more than check if colRefs is non-empty. Example:

```sql
DROP TABLE IF EXISTS t;
DROP TABLE IF EXISTS t2;
DROP ROLE IF EXISTS alice;

CREATE ROLE alice;

CREATE TABLE t (
  k INT PRIMARY KEY,
  a INT
);

INSERT INTO t VALUES (1, 10);

GRANT SELECT, INSERT, UPDATE, DELETE ON t TO alice;

ALTER TABLE t ENABLE ROW LEVEL SECURITY;

CREATE POLICY select_policy_alice
ON t
FOR SELECT
TO alice
USING (a > 0);

CREATE POLICY update_policy_alice
ON t
FOR UPDATE
TO alice
USING (true);

CREATE TABLE t2 (k INT PRIMARY KEY, a INT);

INSERT INTO t2 VALUES (1, -1);

GRANT SELECT, INSERT, UPDATE, DELETE ON t2 TO alice;

SET ROLE alice;

UPDATE t SET a = -2 FROM t2 WHERE true RETURNING t2.k;
-- In Postgres, this update succeeds:
--  k
-- ---
--  1
-- (1 row)
--
-- UPDATE 1
--
-- In CRDB with #144951, the update errors.
-- ERROR: new row violates row-level security policy for table "t"
```
</blockquote></details>

Good catch. I found that checking only `mb.fetchColIDs` wasn’t sufficient. When evaluating the RETURNING clause, `colRefs` is updated with the post-update column values, which makes sense since we want to return the final state—not the values fetched during the scan. So now, I also check the columns in the RETURNING in-scope in addition to `mb.fetchColIDs`.
___
*[`pkg/sql/logictest/testdata/logic_test/row_level_security` line 737 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/145890#-OPBa8N5BIiQKR8uZD1t:-OPGLl7B3701Y_9P39W8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b73b6fce76a314acf40edbfd835a0acb3da8d26f/pkg/sql/logictest/testdata/logic_test/row_level_security#L737)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Can you add a success case where the RETURNING clause has a subquery referencing another table, e.g., `INSERT INTO ins VALUES (7) RETURNING (SELECT k FROM t2);`? It looks like it works fine.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner reviewed (changes requested):
> Reviewed 2 of 10 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145890)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/opt/optbuilder/update.go` line 375 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/145890#-OPVvaES5-Idk-knVyh8:-OPVvaES5-Idk-knVyh9:b-sswjza) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c590dc522136a93a11a016509186ad0ffdcac/pkg/sql/opt/optbuilder/update.go#L375)):*
> ```Go
> 		if !includeSelectPolicies && returningInScope != nil {
> 			for _, sc := range returningInScope.cols {
> 				if sc.table == mb.alias && sc.id != 0 && colRefs.Contains(sc.id) {
> ```

This is too fragile. We shouldn't rely on an alias for something like this. There could be a bug here.

Let me play around with this and find a better way to determine the source of columns referenced in RETURNING.
___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 1699 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/145890#-OPVuU4V2X9D4yfpjMQt:-OPVuU4V2X9D4yfpjMQu:bxjssf4) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c590dc522136a93a11a016509186ad0ffdcac/pkg/sql/opt/optbuilder/mutation_builder.go#L1699)):*
> ```Go
> 	// changed since the scopes were first created.
> 	returningInScope.expr = mb.outScope.expr
> 	returningOutScope.expr = mb.outScope.expr
> ```

I don't think we should be setting `returningOutScope.expr` here. Is there a reason you added this?


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/opt/optbuilder/update.go` line 375 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/145890#-OPVvaES5-Idk-knVyh8:-OPW85Yr-SdwsX0-30qj:bwhjytu) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c590dc522136a93a11a016509186ad0ffdcac/pkg/sql/opt/optbuilder/update.go#L375)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This is too fragile. We shouldn't rely on an alias for something like this. There could be a bug here.

Let me play around with this and find a better way to determine the source of columns referenced in RETURNING.
</blockquote></details>

This should work:

```go
		for i, n := 0, mb.tab.ColumnCount(); i < n; i++ {
			if colRefs.Contains(mb.tabID.ColumnID(i)) {
				includeSelectPolicies = true
				break
			}
		}
```

This checks if any of the columns in `colRefs` reference a column in the target table.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/rls: apply SELECT policies only if RETURNING references columns

Previously, SELECT policies were always applied when a RETURNING clause
was used in an INSERT or UPDATE statement. This behavior was overly
restrictive, as the SELECT policies should only be enforced when the
RETURNING clause actually references one or more columns from the table.

This change modifies the logic to inspect the RETURNING clause and apply
SELECT policies only when column references are present.

Fixes #144951

Epic: CRDB-11724

Release note (bug fix): Fixed incorrect application of SELECT policies
to RETURNING clauses in INSERT and UPDATE when no table columns were
referenced.

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/145890)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @spilchen)

___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 1699 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/145890#-OPVuU4V2X9D4yfpjMQt:-OPWJAJ91bwaLVd2_iUl:b-gdq5ae) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c590dc522136a93a11a016509186ad0ffdcac/pkg/sql/opt/optbuilder/mutation_builder.go#L1699)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I don't think we should be setting `returningOutScope.expr` here. Is there a reason you added this?
</blockquote></details>

I was trying to match the logic in `buildReturning`, which we use to call. I did need to update `expr` in the `returningInScope`, but I mistakenly applied that to `returningOutScope`. After reviewing `buildReturning` again, I see that updating `returningOutScope` wasn't necessary. I'll remove that change.
___
*[`pkg/sql/opt/optbuilder/update.go` line 375 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/145890#-OPVvaES5-Idk-knVyh8:-OPWLAHt8-IgFQCZAb9-:b-73dfxb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e4c590dc522136a93a11a016509186ad0ffdcac/pkg/sql/opt/optbuilder/update.go#L375)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This should work:

```go
		for i, n := 0, mb.tab.ColumnCount(); i < n; i++ {
			if colRefs.Contains(mb.tabID.ColumnID(i)) {
				includeSelectPolicies = true
				break
			}
		}
```

This checks if any of the columns in `colRefs` reference a column in the target table.
</blockquote></details>

Much simpler. Thanks.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
LGTM

spilchen commented:
tftr!

bors r+
